modules/payment/js/credit-card-validation.js

```diff
@@ -0,0 +1,258 @@
+/**
+ * @file
+ * Attaches credit card validation logic.
+ */
+
+(function ($) {
+
+  'use strict';
+
+  var types = {};
+  var VISA = 'visa';
+  var MASTERCARD = 'master-card';
+  var AMEX = 'amex';
+  var DINERSCLUB = 'dinersclub';
+  var DISCOVER = 'discover';
+  var MAESTRO = 'maestro';
+
+  types[VISA] = {
+    niceType: 'Visa',
+    type: VISA,
+    pattern: ['4'],
+    gaps: [4, 8, 12],
+    lengths: [16]
+  };
+
+  types[MASTERCARD] = {
+    niceType: 'MasterCard',
+    type: MASTERCARD,
+    pattern: ['51-55', '222100-272099'],
+    gaps: [4, 8, 12],
+    lengths: [16]
+  };
+
+  types[AMEX] = {
+    niceType: 'American Express',
+    type: AMEX,
+    pattern: ['34', '37'],
+    lengths: [15]
+  };
+
+  types[DINERSCLUB] = {
+    niceType: 'Diners Club',
+    type: DINERSCLUB,
+    pattern: ['300-305', '309', '36', '38', '39'],
+    lengths: [14]
+  };
+
+  types[DISCOVER] = {
+    niceType: 'Discover Card',
+    type: DISCOVER,
+    pattern: ['6011', '622126-622925', '644-649', '65'],
+    lengths: [16, 19]
+  };
+
+  types[MAESTRO] = {
+    niceType: 'Maestro',
+    type: MAESTRO,
+    pattern: [
+      '5018',
+      '5020',
+      '5038',
+      '5612',
+      '5893',
+      '6304',
+      '6759',
+      '6761',
+      '6762',
+      '6763',
+      '0604',
+      '6390'
+    ],
+    lenghts: [12, 13, 14, 15, 16, 17, 18, 19]
+  };
+
+
+  /**
+   * Detect the type of credit card.
+   *
+   * @param {string} number
+   *
+   * @return {object|boolean}
+   */
+  var detectType = function (number) {
+    // Loop over all available types.
```

Comment:
Comments that retell what the code is doing ("Loop over all the available types") aren't useful, let's remove them.